repair/repair.cc

```diff
@@ -1989,26 +2024,81 @@ future<> repair_service::do_rebuild_replace_with_repair(locator::token_metadata_
             }
             auto& strat = erm->get_replication_strategy();
             dht::token_range_vector ranges = strat.get_ranges(myid, *tmptr).get();
-            auto& topology = erm->get_token_metadata().get_topology();
             std::unordered_map<dht::token_range, repair_neighbors> range_sources;
             auto nr_tables = get_nr_tables(db, keyspace_name);
-            rlogger.info("{}: started with keyspace={}, source_dc={}, nr_ranges={}, ignore_nodes={}", op, keyspace_name, source_dc, ranges.size() * nr_tables, ignore_nodes);
+            sstring source_dc_for_keyspace;
+            sstring alt_dc_for_keyspace;
+            auto find_alternative_datacenter = [&] {
+                std::vector<sstring> dcs;
+                dcs.reserve(live_nodes_per_dc.size());
+                std::ranges::copy_if(topology.get_datacenters(), std::back_inserter(dcs), [&] (const auto& dc) {
+                    return dc != *source_dc && !live_nodes_per_dc[dc].empty();
+                });
+                if (!dcs.empty()) {
+                    std::uniform_int_distribution<int> dist(0, dcs.size() - 1);
+                    return dcs[dist(_random_engine)];
+                }
+                return sstring();
+            };
+            // Allow repairing in the source_dc only if there are enough replicas remaining
+            if (source_dc) {
+                switch (strat.get_type()) {
+                case locator::replication_strategy_type::network_topology: {
+                    const auto& nt_strat = dynamic_cast<const locator::network_topology_strategy&>(strat);
+                    size_t rf = nt_strat.get_replication_factor(*source_dc);
+                    auto lost = lost_nodes_per_dc[*source_dc];
+                    source_dc_for_keyspace = *source_dc;
+
+                    // Look for an alternative source_dc if it's unsafe to repair only
+                    // from the source_dc.
+                    if (rf == 0 || lost > 1 || (lost == 1 && rf == 1)) {
+                        alt_dc_for_keyspace = find_alternative_datacenter();
+                        auto msg = format("{}: it is unsafe to use source_dc={} to repair/rebuild keyspace={} since it lost {} nodes, rf={}", op, source_dc, keyspace_name, lost, rf);
+                        if (alt_dc_for_keyspace.empty()) {
+                            msg = fmt::format("{}: found no alternative datacenter", msg);
+                            if (source_dc.force()) {
+                                rlogger.warn("{}: using source_dc anyway according to the force option", msg);
+                            } else if (source_dc.user_provided()) {
+                                throw std::runtime_error(format("{}: Omit the source_dc option to use all nodes in the cluster, or use the --force option to enforce it", msg));
+                            } else {
+                                rlogger.warn("{}: falling back to repair using all nodes", msg);
```

Comment:
how about: falling back to sync data using all replicas.

Reply:
done in v8